NullPointerException  because "this.launcherFuture" is null

Following exception happens on shutdown (very rarely).
```
!ENTRY org.eclipse.lsp4e 4 0 2023-06-07 23:46:45.902
!MESSAGE java.lang.NullPointerException: Cannot invoke "java.util.concurrent.Future.isCancelled()" because "this.launcherFuture" is null
!STACK 0
java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "java.util.concurrent.Future.isCancelled()" because "this.launcherFuture" is null
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$9(LanguageServerWrapper.java:479)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.concurrent.Future.isCancelled()" because "this.launcherFuture" is null
	at org.eclipse.lsp4e.LanguageServerWrapper.isActive(LanguageServerWrapper.java:417)
	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$5(LanguageServerWrapper.java:322)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.request(RemoteEndpoint.java:161)
	at org.eclipse.lsp4j.jsonrpc.services.EndpointProxy.invoke(EndpointProxy.java:91)
	at jdk.proxy10/jdk.proxy10.$Proxy32.shutdown(Unknown Source)
	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$9(LanguageServerWrapper.java:477)
	... 4 more
```

The reason seems to be the lack of synchronization resulting in a change of a prechecked non-null value in the middle of expression computation.

--- a/org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/LanguageServerWrapperTest.java
+++ b/org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/LanguageServerWrapperTest.java
@@ -15,7 +15,11 @@ import static org.eclipse.lsp4e.test.TestUtils.waitForAndAssertCondition;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
+import java.io.IOException;
 import java.util.Collection;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ForkJoinPool;
 
 import org.eclipse.core.resources.IFile;
 import org.eclipse.core.resources.IProject;
@@ -62,5 +66,50 @@ public class LanguageServerWrapperTest {
 
 		TestUtils.closeEditor(editor1, false);
 		TestUtils.closeEditor(editor2, false);
+	}
+	
+	/**
+	 * Check if {@code isActive()} is correctly synchronized with  {@code stop()} 
+	 * @see https://github.com/eclipse/lsp4e/pull/688
+	 */
+	@Test
+	public void testStopAndActive() throws CoreException, IOException, AssertionError, InterruptedException, ExecutionException {
+		IFile testFile1 = TestUtils.createFile(project1, "shouldUseExtension.lsptWithMultiRoot", "");
+		IEditorPart editor1 = TestUtils.openEditor(testFile1);
+		@NonNull Collection<LanguageServerWrapper> wrappers = LanguageServiceAccessor.getLSWrappers(testFile1, request -> true);
+		assertEquals(1, wrappers.size());
+		LanguageServerWrapper wrapper = wrappers.iterator().next();
+		CountDownLatch started = new CountDownLatch(1);
+		try {
+			var startStopJob = ForkJoinPool.commonPool().submit(() -> {
+				started.countDown();
+				while (!Thread.interrupted()) {
+					wrapper.stop();
+					try {
+						wrapper.start();
+					} catch (IOException e) {
+						throw new RuntimeException(e);
+					}
+				}
+			});
+			try {
+				started.await();
+				for (int i = 0; i < 10000000; i++) {
+					// Should not throw
+					wrapper.isActive();
+					if (startStopJob.isDone()) {
+						throw new AssertionError("Job should run indefinitely");
+					}
+				}
+			} finally {
+				startStopJob.cancel(true);
+				if (!startStopJob.isCancelled()) {
+					startStopJob.get();
+				}
+			}
+		} finally {
+			TestUtils.closeEditor(editor1, false);	
+		}
+		
 	}
 }

--- a/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java
+++ b/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java
@@ -389,7 +389,7 @@ public class LanguageServerWrapper {
 	/**
 	 * @return whether the underlying connection to language server is still active
 	 */
-	public boolean isActive() {
+	public synchronized boolean isActive() {
 		return this.launcherFuture != null && !this.launcherFuture.isDone() && !this.launcherFuture.isCancelled();
 	}
 

